### PR TITLE
Sujith/77 impliment events

### LINKED ISSUE #77: General: Implement Events
Each module should emit events for any state-modifying change. Not crucial for the correct functioning of the blockchain but adding events increases overall transparency.

### PR DESCRIPTION
**Pull request by Persistencesujith**
Events implementation

Closes #77

### CODE REVIEW

File: x/did/internal/types/events.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package types
+
+const (
+	EventTypeAddDidDoc     = "create_did_doc"
+	EventTypeAddCredential = "credential"
+
+	AttributeKeyDid        = "did"
+	AttributeKeyPubKey     = "pub_key"
+	AttributeKeyCredType   = "credType"
```
Comment (migueldingli1997) [inline x/did/internal/types/events.go:9]:
```suggestion
	AttributeKeyCredType   = "cred_type"
```

File: x/payments/internal/types/events.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+	AttributeKeyPayerDid          = "payer_did"
+	AttributeKeyPaymentContractId = "payment_contract-id"
+	AttributeKeyAuthorised        = "authorised"
+	AttributeKeyCreatorDid        = "creator_did"
+	AttributeKeyPaymentTemplate   = "payment_template"
+	AttributeKeyPaymentTemplateId = "payment_template_id"
+	AttributeKeyPayer             = "payer"
+	AttributeKeyDeAuthorise       = "can_de_authorise"
+	AttributeKeyDiscountId        = "discountId"
```
Comment (migueldingli1997) [inline x/payments/internal/types/events.go:20]:
```suggestion
	AttributeKeyDiscountId        = "discount_id"
```

File: x/payments/internal/types/events.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	EventTypeEffectPayment                = "effect_payment"
+
+	AttributeKeyPayerDid          = "payer_did"
+	AttributeKeyPaymentContractId = "payment_contract-id"
+	AttributeKeyAuthorised        = "authorised"
+	AttributeKeyCreatorDid        = "creator_did"
+	AttributeKeyPaymentTemplate   = "payment_template"
+	AttributeKeyPaymentTemplateId = "payment_template_id"
+	AttributeKeyPayer             = "payer"
+	AttributeKeyDeAuthorise       = "can_de_authorise"
```
Comment (migueldingli1997) [inline x/payments/internal/types/events.go:19]:
```suggestion
	AttributeKeyDeAuthorise       = "can_deauthorise"
```

File: x/project/internal/types/events.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package types
+
+const (
+	EventTypeCreateProject       = "create_project"
+	EventTypeUpdateProjectStatus = "update_project_status"
+
+	AttributeKeyTxHash       = "txHash"
```
Comment (migueldingli1997) [inline x/project/internal/types/events.go:7]:
```suggestion
	AttributeKeyTxHash       = "tx_hash"
```

File: x/did/internal/types/events.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package types
+
+const (
+	EventTypeAddDidDoc     = "create_did_doc"
```
Comment (migueldingli1997) [inline x/did/internal/types/events.go:4]:
```suggestion
	EventTypeAddDidDoc     = "add_did_doc"
```

File: x/did/internal/types/events.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+	EventTypeAddCredential = "credential"
```
Comment (migueldingli1997) [inline x/did/internal/types/events.go:5]:
```suggestion
	EventTypeAddCredential = "add_credential"
```

File: x/payments/internal/types/events.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package types
+
+const (
+	EventTypePaymentContractAuthorisation = "payment_contract_authorisation"
+	EventTypeCreatePaymentTemplate        = "create_payment_template"
+	EventTypeCreatePaymentContract        = "create_payment_contract"
+	EventTypeCreateSubscription           = "create_subscription"
+	EventTypeGrantDiscount                = "grant-discount"
```
Comment (migueldingli1997) [inline x/payments/internal/types/events.go:8]:
```suggestion
	EventTypeGrantDiscount                = "grant_discount"
```

File: x/treasury/internal/types/events.go [hunk 8]
```diff
@@ -1,5 +1,13 @@
 package types
 
 const (
+	EventTypeMsgSend = "msg_send"
```
Comment (migueldingli1997) [inline x/treasury/internal/types/events.go:4]:
```suggestion
	EventTypeSend = "send"
```

File: x/treasury/handler.go [hunk 9]
```diff
@@ -34,12 +34,18 @@ func handleMsgSend(ctx sdk.Context, k keeper.Keeper, msg types.MsgSend) sdk.Resu
 		return err.Result()
 	}
 
-	ctx.EventManager().EmitEvent(
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeMsgSend,
```
Comment (migueldingli1997) [inline x/treasury/handler.go:39]:
```suggestion
			types.EventTypeSend,
```

File: x/treasury/handler.go [hunk 10]
```diff
@@ -50,12 +56,20 @@ func handleMsgOracleTransfer(ctx sdk.Context, k keeper.Keeper, msg types.MsgOrac
...
```
Comment (migueldingli1997) [inline x/treasury/handler.go:61]: This should be a new event type `EventTypeOracleTransfer`

File: x/treasury/handler.go [hunk 11]
```diff
@@ -66,12 +80,19 @@ func handleMsgOracleMint(ctx sdk.Context, k keeper.Keeper, msg types.MsgOracleMi
...
```
Comment (migueldingli1997) [inline x/treasury/handler.go:85]: This should be a new event type `EventTypeOracleMint`

File: x/treasury/handler.go [hunk 12]
```diff
@@ -82,12 +103,19 @@ func handleMsgOracleBurn(ctx sdk.Context, k keeper.Keeper, msg types.MsgOracleBu
...
```
Comment (migueldingli1997) [inline x/treasury/handler.go:108]: This should be a new event type `EventTypeOracleBurn`

File: x/treasury/internal/types/events.go (hunk 8 above)
Comment (migueldingli1997) [reply x/treasury/internal/types/events.go:4]: Would prefer if we stick to excluding the 'Msg' part

File: x/did/handler.go [hunk 13]
```diff
@@ -27,14 +27,38 @@ func handleMsgAddDidDoc(ctx sdk.Context, k keeper.Keeper, msg types.MsgAddDid) s
...
 	err := k.AddCredentials(ctx, msg.DidCredential.Claim.Id, msg.DidCredential)
 	if err != nil {
 		return err.Result()
 	}
-
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeAddCredential,
+			sdk.NewAttribute(types.AttributeKeyCredType, msg.DidCredential.CredType[0]),
```
Comment (migueldingli1997) [inline x/did/handler.go:53]: Here we want to output the entire slice, not just the first item.

File: x/did/handler.go [hunk 14]
```diff
@@ -27,14 +27,38 @@ func handleMsgAddDidDoc(ctx sdk.Context, k keeper.Keeper, msg types.MsgAddDid) s
...
+			sdk.NewAttribute(types.AttributeKeyIssuer, msg.DidCredential.Issuer),
+			sdk.NewAttribute(types.AttributeKeyIssued, msg.DidCredential.Issued),
+			sdk.NewAttribute(types.AttributeKeyClaimID, msg.DidCredential.Claim.Id),
```
Comment (migueldingli1997) [inline x/did/handler.go:56]: Here we want to output whether the claim is KYC validated `msg.DidCredential.Claim.KYCValidated` not just the claim ID. Please add an attribute for it `AttributeKeyClaimKYCValidated`.

File: x/payments/handler.go [hunk 15]
```diff
@@ -96,7 +96,19 @@ func handleMsgSetPaymentContractAuthorisation(ctx sdk.Context, k Keeper, msg Msg
 		return err.Result()
 	}
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypePaymentContractAuthorisation,
+			sdk.NewAttribute(types.AttributeKeyPayerDid, msg.PayerDid),
+			sdk.NewAttribute(types.AttributeKeyPaymentContractId, msg.PaymentContractId),
```
Comment (migueldingli1997) [inline x/payments/handler.go:103]: Here we are missing the `msg.Authorised` value as an attribute.

File: x/payments/handler.go [hunk 16]
```diff
@@ -129,7 +141,18 @@ func handleMsgCreatePaymentTemplate(ctx sdk.Context, k Keeper, bk bank.Keeper, m
 	// Submit payment template
 	k.SetPaymentTemplate(ctx, msg.PaymentTemplate)
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreatePaymentTemplate,
+			sdk.NewAttribute(types.AttributeKeyCreatorDid, msg.CreatorDid),
```
Comment (migueldingli1997) [inline x/payments/handler.go:147]: Here we want to output all attributes of the payment template, not just the creator DID.

File: x/payments/handler.go [hunk 17]
```diff
@@ -175,7 +198,22 @@ func handleMsgCreatePaymentContract(ctx sdk.Context, k Keeper, bk bank.Keeper,
 	// Submit payment contract
 	k.SetPaymentContract(ctx, contract)
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
```
Comment (migueldingli1997) [inline x/payments/handler.go:202]: We are missing the CanDeauthorise attribute here.

File: x/payments/handler.go [hunk 18]
```diff
@@ -221,7 +259,21 @@ func handleMsgCreateSubscription(ctx sdk.Context, k Keeper,
 	// Submit subscription
 	k.SetSubscription(ctx, subscription)
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
```
Comment (migueldingli1997) [inline x/payments/handler.go:263]: We are missing the creator DID as an event attribute here.

File: x/payments/handler.go [hunk 19]
```diff
@@ -221,7 +259,21 @@ func handleMsgCreateSubscription(ctx sdk.Context, k Keeper,
...
+			types.EventTypeCreateSubscription,
+			sdk.NewAttribute(types.AttributeKeySubscriptionId, msg.SubscriptionId),
+			sdk.NewAttribute(types.AttributeKeyPaymentContractId, msg.PaymentContractId),
+			sdk.NewAttribute(types.AttributeKeyMaxPeriods, msg.MaxPeriods.String()),
+			sdk.NewAttribute(types.AttributeKeyPeriod, msg.Period.GetPeriodUnit()),
```
Comment (migueldingli1997) [inline x/payments/handler.go:268]: Please change the attribute key to `AttributeKeyPeriodUnit` (=`"period_unit"`).

File: x/project/handler.go [hunk 20]
```diff
@@ -227,7 +268,21 @@ func handleMsgCreateClaim(ctx sdk.Context, k Keeper, fk payments.Keeper,
...
 	}
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUpdateProjectStatus,
+			sdk.NewAttribute(types.AttributeKeyTxHash, msg.TxHash),
+			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
+			sdk.NewAttribute(types.AttributeKeyProjectDid, msg.ProjectDid),
+			sdk.NewAttribute(types.AttributeKeyData, msg.Data.ClaimID),
```
Comment (migueldingli1997) [inline x/project/handler.go:277]: Please add `msg.Data.Status` as an attribute. And then the attribute keys would be: `AttributeKeyClaimID` and `AttributeKeyClaimStatus`.

File: x/project/handler.go [hunk 21]
```diff
@@ -120,7 +133,22 @@ func handleMsgUpdateProjectStatus(ctx sdk.Context, k Keeper, bk bank.Keeper,
...
 	k.SetProjectDoc(ctx, projectDoc)
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUpdateProjectStatus,
+			sdk.NewAttribute(types.AttributeKeyTxHash, msg.TxHash),
+			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
+			sdk.NewAttribute(types.AttributeKeyProjectDid, msg.ProjectDid),
+			sdk.NewAttribute(types.AttributeKeyData, msg.Data.EthFundingTxnID),
```
Comment (migueldingli1997) [inline x/project/handler.go:142]: Attribute key should match the data that we're outputting. So for `EthFundingTxnID`, the key should be `AttributeKeyEthFundingTxnID`.

Comment (migueldingli1997) [reply x/project/handler.go:142]: Additionally, we are missing the new status as an attribute value. `AttributeKeyStatus`

File: x/project/handler.go [hunk 22]
```diff
@@ -194,8 +222,21 @@ func handleMsgCreateAgent(ctx sdk.Context, k Keeper, bk bank.Keeper, msg MsgCrea
...
 	}
-
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUpdateProjectStatus,
+			sdk.NewAttribute(types.AttributeKeyTxHash, msg.TxHash),
+			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
+			sdk.NewAttribute(types.AttributeKeyProjectDid, msg.ProjectDid),
+			sdk.NewAttribute(types.AttributeKeyData, msg.Data.AgentDid),
```
Comment (migueldingli1997) [inline x/project/handler.go:231]: Please add `msg.Data.Role` as an attribute. And then the attribute keys would be: `AttributeKeyAgentDid` and `AttributeKeyAgentRole`.

File: x/project/handler.go [hunk 23]
```diff
@@ -252,7 +307,21 @@ func handleMsgCreateEvaluation(ctx sdk.Context, k Keeper, fk payments.Keeper, bk
...
 	}
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUpdateProjectStatus,
+			sdk.NewAttribute(types.AttributeKeyTxHash, msg.TxHash),
+			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
+			sdk.NewAttribute(types.AttributeKeyProjectDid, msg.ProjectDid),
+			sdk.NewAttribute(types.AttributeKeyData, msg.Data.ClaimID),
```
Comment (migueldingli1997) [inline x/project/handler.go:316]: Missing claim status as attribute value.

File: x/project/handler.go [hunk 24]
```diff
@@ -290,7 +359,21 @@ func handleMsgWithdrawFunds(ctx sdk.Context, k Keeper, bk bank.Keeper,
...
 	}
 
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUpdateProjectStatus,
+			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
+			sdk.NewAttribute(types.AttributeKeyData, msg.Data.ProjectDid),
+			sdk.NewAttribute(types.AttributeKeyRecipientDid, msg.Data.RecipientDid),
+			sdk.NewAttribute(types.AttributeKeyAmount, msg.Data.Amount.String()),
```
Comment (migueldingli1997) [inline x/project/handler.go:368]: Missing attribute value `msg.Data.IsRefund`

File: x/project/handler.go (hunk 20 above)
Comment (Persistencesujith) [reply x/project/handler.go]: AttributeKeyClaimStatus did not find..
type CreateClaimDoc struct {
	ClaimID string `json:"claimID" yaml:"claimID"`
}

File: x/payments/handler.go [hunk 25]
```diff
@@ -175,7 +206,23 @@ func handleMsgCreatePaymentContract(ctx sdk.Context, k Keeper, bk bank.Keeper,
...
-	return sdk.Result{}
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreatePaymentContract,
+			sdk.NewAttribute(types.AttributeKeyCreatorDid, msg.CreatorDid),
+			sdk.NewAttribute(types.AttributeKeyPaymentTemplateId, msg.PaymentTemplateId),
+			sdk.NewAttribute(types.AttributeKeyPaymentContractId, msg.PaymentContractId),
+			sdk.NewAttribute(types.AttributeKeyPayer, msg.Payer.String()),
+			sdk.NewAttribute(types.AttributeKeyDiscountId, msg.DiscountId.String()),
+			sdk.NewAttribute(types.AttributeKeyDeAuthorise, strconv.FormatBool(msg.CanDeauthorise)),
```
Comment (migueldingli1997) [inline x/payments/handler.go:217]: Attribute key should match the value we're outputting, so it should be `AttributeKeyCanDeauthorise`

File: x/did/internal/types/events.go (hunk 1 above)
Comment (Persistencesujith) [reply x/did/internal/types/events.go]: AttributeKeyCredType   = "credType" had been changed to AttributeKeyCredType     = "cred_type" . hope this is fine

File: x/payments/internal/types/events.go (hunk 2 above)
Comment (Persistencesujith) [reply x/payments/internal/types/events.go]: AttributeKeyDiscountId        = "discountId" has been added as AttributeKeyDiscountId         = "discount_id".

File: x/payments/internal/types/events.go (hunk 3 above)
Comment (Persistencesujith) [reply x/payments/internal/types/events.go]: changed to AttributeKeyCanDeauthorise     = "can_deauthorise"

File: x/project/internal/types/events.go (hunk 4 above)
Comment (Persistencesujith) [reply x/project/internal/types/events.go]: code updated as per suggested AttributeKeyTxHash       = "tx_hash"

File: x/did/internal/types/events.go (hunk 5 above)
Comment (Persistencesujith) [reply x/did/internal/types/events.go]: code updated from EventTypeAddDidDoc     = "create_did_doc" to  EventTypeAddDidDoc     = "add_did_doc"

File: x/did/internal/types/events.go (hunk 6 above)
Comment (Persistencesujith) [reply x/did/internal/types/events.go]: code has updated from EventTypeAddCredential = "add_credential" TO EventTypeAddCredential = "add_credential"

File: x/payments/internal/types/events.go (hunk 7 above)
Comment (Persistencesujith) [reply x/payments/internal/types/events.go]: code has been added from EventTypeGrantDiscount                = "grant-discount" to  EventTypeGrantDiscount                = "grant_discount"

File: x/treasury/internal/types/events.go (hunk 8 above)
Comment (Persistencesujith) [reply x/treasury/internal/types/events.go]: Ok. updated the code to EventTypeSend  = "send"

File: x/treasury/handler.go (hunk 9 above)
Comment (Persistencesujith) [reply x/treasury/handler.go]: ok. updated the code types.EventTypeSend

File: x/treasury/handler.go (hunk 10 above)
Comment (Persistencesujith) [reply x/treasury/handler.go]: ok. Updated the code 

ctx.EventManager().EmitEvents(sdk.Events{
		sdk.NewEvent(
			types.EventTypeOracleTransfer,
			sdk.NewAttribute(types.AttributeKeyOracleDid, msg.OracleDid),
			sdk.NewAttribute(types.AttributeKeyTxHashFromDid, msg.FromDid),
			sdk.NewAttribute(types.AttributeKeyToDidOrAddr, msg.ToDidOrAddr),
			sdk.NewAttribute(types.AttributeKeyAmount, msg.Amount.String()),
			sdk.NewAttribute(types.AttributeKeyProof, msg.Proof),
		),

File: x/treasury/handler.go (hunk 11 above)
Comment (Persistencesujith) [reply x/treasury/handler.go]: ok. Modified the code accordingly

File: x/treasury/handler.go (hunk 12 above)
Comment (Persistencesujith) [reply x/treasury/handler.go]: ok. Modified the code accordingly

File: x/did/handler.go (hunk 13 above)
Comment (Persistencesujith) [reply x/did/handler.go]: ok. Noted and added in the similar way

File: x/project/handler.go (hunk 21 above)
Comment (Persistencesujith) [reply x/project/handler.go]: both has been updated .
sdk.NewAttribute(types.AttributeKeyEthFundingTxnID, msg.Data.EthFundingTxnID),
			sdk.NewAttribute(types.AttributeKeyStatus, fmt.Sprint(msg.Data.Status)),

File: x/project/handler.go (hunk 22 above)
Comment (Persistencesujith) [reply x/project/handler.go]: ok. missed to update in the first MR review on AttributeKeyAgentRole.
will update the code and push

File: x/project/handler.go (hunk 23 above)
Comment (Persistencesujith) [reply x/project/handler.go]: status has been added

File: x/payments/handler.go (hunk 25 above)
Comment (Persistencesujith) [reply x/payments/handler.go]: resolved the naming of AttributeKeyCanDeauthorise , AttributeKeyAgentRole

File: x/payments/handler.go (hunk 17 above)
Comment (Persistencesujith) [reply x/payments/handler.go:210]: Ok . Added the key

File: x/payments/handler.go (hunk 16 above)
Comment (Persistencesujith) [reply x/payments/handler.go:149]: Ok . Added payment template

File: x/payments/handler.go (hunk 15 above)
Comment (Persistencesujith) [reply x/payments/handler.go:104]: Ok. Added the msg.Authorised

File: x/did/handler.go (hunk 14 above)
Comment (Persistencesujith) [reply x/did/handler.go:58]: added the AttributeKeyClaimKYCValidated

File: x/payments/internal/types/msgs.go [hunk 26]
```diff
@@ -314,8 +314,12 @@ func (msg MsgRevokeDiscount) GetSignBytes() []byte {
...
 
 type MsgEffectPayment struct {
-	SenderDid         did.Did `json:"sender_did" yaml:"sender_did"`
-	PaymentContractId string  `json:"payment_contract_id" yaml:"payment_contract_id"`
+	SenderDid                 did.Did `json:"sender_did" yaml:"sender_did"`
+	PaymentContractId         string  `json:"payment_contract_id" yaml:"payment_contract_id"`
+	InputFromPayRemainderPool string  `json:"input_from_pay_remainder_pool" yaml:"input_from_pay_remainder_pool`
+	InputFromPayer            string  `json:"input_from_payer" yaml:"input_from_payer"`
+	OutputToPayRemainderPool  string  `json:"output_to_pay_remainder_pool" yaml:"output_to_pay_remainder_pool"`
+	OutputToPayees            string  `json:"output_to_payees" yaml:"output_to_payees"`
```
Comment (migueldingli1997) [inline x/payments/internal/types/msgs.go:322]: We don't need to change the message. Please revert these changes.

File: x/payments/handler.go [hunk 27]
```diff
@@ -416,6 +416,10 @@ func handleMsgEffectPayment(ctx sdk.Context, k Keeper, bk bank.Keeper, msg MsgEf
 			types.EventTypeEffectPayment,
 			sdk.NewAttribute(types.AttributeKeySenderDid, msg.SenderDid),
 			sdk.NewAttribute(types.AttributeKeyPaymentContractId, msg.PaymentContractId),
+			sdk.NewAttribute(types.AttributeKeyInputFromPayRemainderPool, msg.OutputToPayRemainderPool),
+			sdk.NewAttribute(types.AttributeKeyInputFromPayer, msg.InputFromPayer),
+			sdk.NewAttribute(types.AttributeKeyOutputToPayRemainderPool, msg.InputFromPayRemainderPool),
+			sdk.NewAttribute(types.AttributeKeyOutputToPayees, msg.OutputToPayees),
```
Comment (migueldingli1997) [inline x/payments/handler.go:422]: We did not need to make these changes. Please revert.

File: x/payments/internal/types/msgs.go (hunk 26 above)
Comment (Persistencesujith) [reply x/payments/internal/types/msgs.go]: removed

File: x/payments/handler.go (hunk 27 above)
Comment (Persistencesujith) [reply x/payments/handler.go]: removed

File: x/payments/internal/keeper/payments.go [hunk 28]
```diff
@@ -1,6 +1,7 @@
 package keeper
 
 import (
+	"fmt"
 	sdk "github.com/cosmos/cosmos-sdk/types"
```
Comment (migueldingli1997) [inline x/payments/internal/keeper/payments.go:5]:
```suggestion
	sdk "github.com/cosmos/cosmos-sdk/types"
```
Unused import
